pgjdbc/src/main/java/org/postgresql/core/v3/ScramAuthenticator.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2017, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.core.v3;
+
+import org.postgresql.core.PGStream;
+import org.postgresql.util.GT;
+import org.postgresql.util.PSQLException;
+import org.postgresql.util.PSQLState;
+
+import com.ongres.scram.client.ScramClient;
+import com.ongres.scram.common.ClientFinalMessage;
+import com.ongres.scram.common.ClientFirstMessage;
+import com.ongres.scram.common.StringPreparation;
+import com.ongres.scram.common.exception.ScramException;
+import com.ongres.scram.common.util.TlsServerEndpoint;
+
+import java.io.IOException;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.SSLSocket;
+
+final class ScramAuthenticator {
+  private static final Logger LOGGER = Logger.getLogger(ScramAuthenticator.class.getName());
+
+  private final PGStream pgStream;
+  private final ScramClient scramClient;
+
+  ScramAuthenticator(char[] password, PGStream pgStream, Properties info) throws PSQLException {
+    this.pgStream = pgStream;
+    this.scramClient = initializeScramClient(password, pgStream, info);
+  }
+
+  private static ScramClient initializeScramClient(char[] password, PGStream stream, Properties info) throws PSQLException {
+    try {
+      final ChannelBindingOption channelBinding = ChannelBindingOption.of(info);
+      LOGGER.log(Level.FINEST, "channelBinding( {0} )", channelBinding);
```

Comment:
use a supplier to avoid having to test for Level.FINEST ?

Reply:
Not needed since the value is not evaluated until matches the level.